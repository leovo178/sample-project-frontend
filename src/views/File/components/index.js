export { default as FetchImage } from './FetchImage'
