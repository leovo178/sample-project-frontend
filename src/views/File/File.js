import React from 'react'

import { useQuery, useMutation } from '@apollo/react-hooks'

import { Box, Button, Grid, Typography } from '@material-ui/core'
import { makeStyles } from '@material-ui/styles'

import { FetchImage } from './components'

import { GET_USER_INFO, GET_FILE } from './gql/query'
import { UPLOAD_FILE, SET_UPLOADED_FILE } from './gql/mutation'

const useStyle = makeStyles(theme => ({
	root: {
		width: 'calc(100% - 80px)',
		height: '100vh',
		padding: theme.spacing(3),
		position: 'relative',
	},
	container: {
		border: `1px solid ${theme.palette.common.border}`,
		height: '100%',
	},
	container__uploader__imageviewer: {
		height: '100%',
	},
	item__uploader: {
		width: 462,
		position: 'relative',
		display: 'flex',
		alignItems: 'center',
	},
	item__uploader__button: {
		fontSize: '0.875rem',
		color: theme.palette.common.white,
		fontWeight: 700,
		textTransform: 'unset',
		marginRight: theme.spacing(1),
		padding: '3px 12px 3px 11px',
	},
	item__uploader__input: {
		position: 'absolute',
		top: 0,
		left: 0,
		opacity: 0,
		height: '100%',
		width: '100%',
	},
	item__uploader__filename: {
		width: 360,
		whiteSpace: 'nowrap',
		textOverflow: 'ellipsis',
		overflow: 'hidden',
	},
	item__imageviewer: {
		width: 462,
		height: 324,
		border: `1px solid ${theme.palette.common.border}`,
		background: theme.palette.common.greyButton,
		display: 'flex',
		justifyContent: 'center',
		alignItems: 'center',
		marginTop: theme.spacing(3),
		overflow: 'hidden',
	},
	item__imageviewer__image: {
		height: '100%',
	},
}))

const File = () => {
	const classes = useStyle()
	const {
		data: { file },
	} = useQuery(GET_FILE)
	const [setUploadedFile] = useMutation(SET_UPLOADED_FILE, {
		onError: err => alert(err),
	})

	const { data } = useQuery(GET_USER_INFO, {
		onCompleted: data => {
			if (data && data.me && data.me.image && !file.filename) {
				// when reload page
				// prevent GET_USER_INFO query use its latest cache when user log out
				setUploadedFile({ variables: { file: data.me.image } })
			}
		},
		onError: err => alert(err),
	})

	const [uploadFile] = useMutation(UPLOAD_FILE, {
		onCompleted: ({ uploadImage }) => {
			setUploadedFile({ variables: { file: uploadImage } })
		},
		onError: err => alert(err),
	})

	const onUploadFile = ({ target }) => {
		const file = target.files[0]

		uploadFile({ variables: { file } })
	}

	return (
		<Box className={classes.root}>
			<Box className={classes.container}>
				<Grid
					container
					direction='column'
					alignItems='center'
					justify='center'
					className={classes.container__uploader__imageviewer}
				>
					<Grid item className={classes.item__uploader}>
						<Button
							variant='contained'
							size='small'
							className={classes.item__uploader__button}
						>
							file select
							<input
								accept='image/*'
								className={classes.item__uploader__input}
								onChange={onUploadFile}
								type='file'
							/>
						</Button>
						<Typography
							variant='body2'
							className={classes.item__uploader__filename}
						>
							{file.filename ? file.filename : 'No file selected'}
						</Typography>
					</Grid>
					<Grid item className={classes.item__imageviewer}>
						{file.link ? (
							<FetchImage
								fileName={file.filename}
								fileLink={file.link}
								styles={classes.item__imageviewer__image}
							/>
						) : (
							<Typography variant='body2'>image</Typography>
						)}
					</Grid>
				</Grid>
			</Box>
		</Box>
	)
}

export default File
