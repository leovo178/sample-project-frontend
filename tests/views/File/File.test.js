import React from 'react'

import waait from 'waait'
import { act, fireEvent } from '@testing-library/react'

import File from '@views/File'

import { findDOMNodeOfFile, fileName, mockFile } from './helpers'

describe('<File />', () => {
	let component

	beforeEach(async () => {
		await act(async () => {
			component = findDOMNodeOfFile(<File />)
		})
	})
	it('should match snapshot', async () => {
		expect(component.container).toMatchSnapshot()
	})

	it('should upload file and show file name correctly', async () => {
		await waait(10)
		global.URL.createObjectURL = jest.fn()

		const inputElement = component.getByTestId('file-input')

		await act(async () => {
			fireEvent.change(inputElement, { target: { files: [mockFile] } })
			fireEvent.input(inputElement)
		})

		expect(component.queryByText(fileName)).toBeTruthy()
	})
})
