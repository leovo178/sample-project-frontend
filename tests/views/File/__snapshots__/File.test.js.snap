// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<File /> should match snapshot 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-11 makeStyles-root-1"
  >
    <div
      class="MuiBox-root MuiBox-root-12 makeStyles-container-2"
    >
      <div
        class="MuiGrid-root makeStyles-container__uploader__imageviewer-3 MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
      >
        <div
          class="MuiGrid-root makeStyles-item__uploader-4 MuiGrid-item"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-item__uploader__button-5 MuiButton-containedSizeSmall MuiButton-sizeSmall"
            data-testid="file-selectbutton"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              file select
              <input
                accept="image/*"
                class="makeStyles-item__uploader__input-6"
                data-testid="file-input"
                type="file"
              />
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <p
            class="MuiTypography-root makeStyles-item__uploader__filename-7 MuiTypography-body2"
          >
            mockimage.jpg
          </p>
        </div>
        <div
          class="MuiGrid-root makeStyles-item__imageviewer-8 MuiGrid-item"
        >
          <p
            class="MuiTypography-root MuiTypography-body2"
          >
            image
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
